src/braket/devices/local_simulator.py

```diff
@@ -331,54 +291,59 @@ def _(
                 source=program.source,
                 inputs=inputs_copy,
             )
+        return (program,)
 
-        results = simulator.run(program, shots, *args, **kwargs)
-
-        if isinstance(results, GateModelQuantumTaskResult):
-            return results
-
-        return GateModelQuantumTaskResult.from_object(results)
-
-    @_run_internal.register
-    def _(
-        self,
-        program: SerializableProgram,
-        shots: Optional[int] = None,
-        inputs: Optional[dict[str, float]] = None,
-        *args,
-        **kwargs,
-    ):
-        program = OpenQASMProgram(source=program.to_ir(ir_type=IRType.OPENQASM))
-        return self._run_internal(program, shots, inputs, *args, **kwargs)
+    @_construct_payload.register
+    def _(self, program: SerializableProgram, _inputs, _shots):
+        return (OpenQASMProgram(source=program.to_ir(ir_type=IRType.OPENQASM)),)
 
-    @_run_internal.register
-    def _(
-        self,
-        program: AnalogHamiltonianSimulation,
-        shots: Optional[int] = None,
-        *args,
-        **kwargs,
-    ):
+    @_construct_payload.register
+    def _(self, program: AnalogHamiltonianSimulation, _inputs, _shots):
         simulator = self._delegate
         if DeviceActionType.AHS not in simulator.properties.action:
             raise NotImplementedError(
                 f"{type(simulator)} does not support analog Hamiltonian simulation programs"
             )
-        results = simulator.run(program.to_ir(), shots, *args, **kwargs)
-        return AnalogHamiltonianSimulationQuantumTaskResult.from_object(results)
+        return (program.to_ir(),)
 
-    @_run_internal.register
-    def _(
-        self,
-        program: AHSProgram,
-        shots: Optional[int] = None,
-        *args,
-        **kwargs,
-    ):
+    @_construct_payload.register
+    def _(self, program: AHSProgram, _inputs, _shots):
         simulator = self._delegate
         if DeviceActionType.AHS not in simulator.properties.action:
             raise NotImplementedError(
                 f"{type(simulator)} does not support analog Hamiltonian simulation programs"
             )
-        results = simulator.run(program, shots, *args, **kwargs)
-        return AnalogHamiltonianSimulationQuantumTaskResult.from_object(results)
+        return (program,)
+
+    @_construct_payload.register
+    def _(self, problem: Problem, _inputs, _shots) -> Union[
+        GateModelQuantumTaskResult,
+        AnalogHamiltonianSimulationQuantumTaskResult,
+        AnnealingQuantumTaskResult,
+    ]:
+        simulator = self._delegate
+        if DeviceActionType.ANNEALING not in simulator.properties.action:
+            raise NotImplementedError(
+                f"{type(simulator)} does not support quantum annealing problems"
+            )
+        return (problem.to_ir(),)
+
+    @singledispatchmethod
+    def _to_result_object(self, result: Any) -> Any:
```

Comment:
more specific `Union` type hint for `result`? I'm asking mainly because the more specific type hints would give better linting/IDE hints when calling these methods. But I recognize it makes the code uglier for a small benefit, so feel free to ignore.